app/src/main/java/org/wikipedia/util/UriUtil.kt

```diff
@@ -0,0 +1,147 @@
+        val chooserIntent = ShareUtil.createChooserIntent(Intent(Intent.ACTION_VIEW, uri), context)
+        try {
+            chooserIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            context.startActivity(chooserIntent)
+        } catch (e: ActivityNotFoundException) {
+            // This means that there was no way to handle this link.
+            // We will just show a toast now. FIXME: Make this more visible?
+            ShareUtil.showUnresolvableIntentMessage(context)
+        }
+    }
+
+    @JvmStatic
+    fun resolveProtocolRelativeUrl(wiki: WikiSite,
+                                   url: String): String {
+        val ret = resolveProtocolRelativeUrl(url)
+
+        // also handle images like /w/extensions/ImageMap/desc-20.png?15600 on Estados Unidos
+        // or like /api/rest_v1/page/graph/png/API/0/019dd76b5f4887040716e65de53802c5033cb40c.png
+        return if (ret.startsWith("./") || ret.startsWith("/w/") ||
+                ret.startsWith("/wiki/") || ret.startsWith("/api/"))
+            wiki.uri().buildUpon().appendEncodedPath(ret.replaceFirst("/".toRegex(), ""))
+                    .build().toString()
+        else ret
+    }
+
+    @JvmStatic
+    fun resolveProtocolRelativeUrl(url: String): String {
+        return if (url.startsWith("//")) WikipediaApp.getInstance().wikiSite.scheme() + ":" + url else url
+    }
+
+    @JvmStatic
+    fun isValidPageLink(uri: Uri): Boolean {
+        return ((!uri.authority.isNullOrEmpty() &&
+                uri.authority!!.endsWith("wikipedia.org") &&
+                !uri.path.isNullOrEmpty() &&
+                uri.path!!.matches(("^$WIKI_REGEX.*").toRegex())) &&
+                (uri.fragment == null || (uri.fragment!!.isNotEmpty() &&
+                        !uri.fragment!!.startsWith("cite"))))
+    }
+
+    @JvmStatic
+    fun handleExternalLink(context: Context, uri: Uri) {
+        visitInExternalBrowser(context, uri)
+    }
+
+    @JvmStatic
+    fun getUrlWithProvenance(context: Context, title: PageTitle,
+                             @StringRes provId: Int): String {
+        return title.uri + "?wprov=" + context.getString(provId)
+    }
+
+    @JvmStatic
+    fun getFilenameFromUploadUrl(url: String): String {
+        val splitList = url.split("/")
+        val thumbnailName = splitList[splitList.size - 1]
+        return if (url.contains("/thumb/") && splitList.size > 2) {
+            splitList[splitList.size - 2]
+        } else thumbnailName
+    }
+
+    @JvmStatic
+    fun getTitleFromUrl(url: String): String {
+        return removeFragment(removeLinkPrefix(url)).replace("_", " ")
+    }
+
+    /** Get language variant code from a Uri, e.g. "zh.*", otherwise returns empty string.  */
+    @JvmStatic
+    fun getLanguageVariantFromUri(uri: Uri): String {
+        if (uri.path.isNullOrEmpty()) {
+            return ""
+        }
+        val parts = StringUtils.defaultString(uri.path).split('/')
```

Comment:
Since we already have the `isNullOrEmpty()`, do we still need this `StringUtils.defaultString()` instead of using `uri.path!!`?